src/Form/index.tsx

```diff
@@ -455,25 +476,92 @@ function Form({
       setActiveStep(JSON.parse(JSON.stringify(curStep)));
     };
 
-  const runUserCallback = async (
-    userCallback: any,
+  const eventCallbackMap: Record<string, any> = {
+    change: onChange,
+    load: onLoad,
+    form_complete: onFormComplete,
+    submit: onSubmit,
+    error: onError,
+    view: onView,
+    action: onAction
+  };
+
+  const eventHasUserLogic = (event: string) => {
+    return (
+      typeof eventCallbackMap[event] === 'function' ||
+      (logicRules &&
+        logicRules.some(
+          (logicRule: LogicRule) => logicRule.trigger_event === event
+        ))
+    );
+  };
+
+  const runUserLogic = async (
+    event: string,
     getProps: () => Record<string, any> = () => ({})
   ) => {
-    if (typeof userCallback !== 'function') return;
-    try {
-      await userCallback({
-        ...getFormContext(_internalId),
-        ...getProps()
-      });
-    } catch (e) {
-      console.warn(e);
+    const props = {
+      ...getFormContext(_internalId),
+      ...getProps()
+    };
+    let logicRan = false;
+    if (typeof eventCallbackMap[event] === 'function') {
+      logicRan = true;
+      await eventCallbackMap[event](props);
     }
+    // filter the logicRules that have trigger_event matching the trigger event (type_)
+    if (logicRules) {
+      const logicRulesForEvent = logicRules.filter(
+        (logicRule: any) => logicRule.trigger_event === event
+      );
+      // Run the logic rules in sequence!
+      for (const logicRule of logicRulesForEvent) {
+        // don't run disabled or invalid rules
+        if (!logicRule.enabled || !logicRule.valid || !logicRule.code) continue;
```

Comment:
we should filter out disabled and invalid rules on the backend so we don't need this check or these flags here